Feature create user

Roman4412 opened the pull request.

RomanPustovalov committed:
feat: add HibernateConfig

added a HibernateConfig.java to get the session factory bean, and hibernate config file

RomanPustovalov committed:
chore: add dependencies

- mapstruct
- hibernate core
- hibernate validator

RomanPustovalov committed:
feat: add user entity

RomanPustovalov committed:
chore: add dao

  - UserDao interface, and hibernate implementation of this interface
  - AbstractSessionTransactionManager to manage sessions and transactions

RomanPustovalov committed:
feat: add a UserService and implement a method to save the user

RomanPustovalov committed:
feat: add UserController

A UserController has been added and methods for registering and displaying the page have been implemented

RomanPustovalov committed:
feat: add static and templates for registration page

RomanPustovalov committed:
chore: change the dependency for the validator

RomanPustovalov committed:
fix: fix the absolute url of the form to a relative one

RomanPustovalov committed:
feat: add validation of client data during user registration

ibobrov commented on src/main/java/com/pustovalov/weatherapplication/config/HibernateConfig.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Bean(destroyMethod = "close")
+    public SessionFactory sessionFactory() {
+        final StandardServiceRegistry registry = new StandardServiceRegistryBuilder().configure().build();
```
Можно оставить 2 строки и не делать try-catch. ПРоблема появится на старте, сам спринг прекрасно обьянит что не так - своии exception не особо нужны

ibobrov commented on src/main/java/com/pustovalov/weatherapplication/dao/AbstractSessionTransactionManager.java:
```diff
@@ -0,0 +1,36 @@
+
+    protected final SessionFactory sessionFactory;
+
+    public <R> R executeInTransaction(Function<Session, R> action) {
```
почему метод public, если ты его наследовать будешь?

ibobrov commented on src/main/java/com/pustovalov/weatherapplication/exception/ObjectAlreadyExistException.java:
```diff
@@ -0,0 +1,8 @@
+
+public class ObjectAlreadyExistException extends RuntimeException {
+
+    public ObjectAlreadyExistException(String message) {
```
лучше сразу все конструкторы из runtime exception имплементи

ibobrov commented on src/main/java/com/pustovalov/weatherapplication/service/UserService.java:
```diff
@@ -0,0 +1,37 @@
+
+        User user = mapper.toEntity(createUserFormData);
+
+        if (userDao.isExist(user)) {
```
некорректный подход - не надо чекать узера, надо от exception отталкиваться (пока ты чекал, пользователя могли добавить)

ibobrov replied:
и лучше делать такое на if(exist) {
  trow...
}

а дальше просто логика

ibobrov commented on src/main/java/com/pustovalov/weatherapplication/service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,16 @@
+@Mapper
+public interface UserMapper {
+
+    UserMapper mapper = Mappers.getMapper(UserMapper.class);
```
в спринге чуть по другому принято, выше написал что надо

ibobrov commented on src/main/java/com/pustovalov/weatherapplication/service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,16 @@
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+@Mapper
```
@Mapper(componentModel = MappingConstants.ComponentModel.SPRING)

Roman4412 reviewed (commented):
1

Roman4412 reviewed (commented):
2

ibobrov commented on src/main/java/com/pustovalov/weatherapplication/dto/CreateUserFormData.java:
```diff
@@ -0,0 +1,16 @@
+
+public record CreateUserFormData(
+        @NotNull(message = "The login field must be filled in")
+        @Size(min = 1 , max = 20, message = "The length of the login must be no more than 20 and no less than 1 character")
```
добавить регулярку

RomanPustovalov committed:
feat: finalize the validation

  - moved validation error messages to messages.properties
  - fixed an error where thymeleaf did not see the transferred object
  - added error message output

RomanPustovalov committed:
fix: fix issues raised in review
